File: artifacts/api-server/src/lib/googleAuth.ts
import { createPublicKey, verify as verifySignature, type KeyObject } from 'node:crypto';
import type { NextFunction, Request, Response } from 'express';
import { logger } from './logger';

/**
 * Server-side verification of a Google ID token.
 *
 * The browser reads the same token to draw a name and an avatar, and is right
 * not to verify it — it has no way to. Nothing on the client side changes here.
 * What changes is that a token now also travels to this server when it is being
 * used to decide something, and here it is checked properly: signature against
 * Google's published keys, issuer, audience, expiry, and a verified email.
 * Anything less and "signed in as" is a claim the caller makes about itself.
 *
 * No library: RS256 is a signature this runtime already verifies, and Google's
 * key set is one well-known JSON document. What that buys is that the rules
 * below are readable in one place rather than configured somewhere.
 */

const JWKS_URL = 'https://www.googleapis.com/oauth2/v3/certs';
const VALID_ISSUERS = new Set(['accounts.google.com', 'https://accounts.google.com']);

/** Tolerance for a client clock that disagrees with Google's, in seconds. */
const CLOCK_SKEW_SECONDS = 60;

/** Floor on how long a fetched key set is reused, when Google says nothing. */
const DEFAULT_JWKS_TTL_MS = 60 * 60 * 1000;

export interface VerifiedGoogleUser {
  /** The `sub` claim: stable per account, and what records should key on. */
  subject: string;
  email: string;
  name: string;
  /** Epoch milliseconds at which the token stops being acceptable. */
  expiresAt: number;
}

interface JsonWebKey {
  kid?: string;
  kty?: string;
  alg?: string;
  use?: string;
  n?: string;
  e?: string;
}

interface CachedKeys {
  keys: Map<string, KeyObject>;
  expiresAt: number;
}

let cache: CachedKeys | undefined;
let inFlight: Promise<CachedKeys> | undefined;

export function googleClientId(): string | undefined {
  const id = process.env.GOOGLE_CLIENT_ID?.trim();
  return id ? id : undefined;
}

/** Whether sign-in-backed routes can work at all on this deployment. */
export function isGoogleAuthConfigured(): boolean {
  return googleClientId() !== undefined;
}

function maxAgeMs(cacheControl: string | null): number {
  const match = cacheControl && /max-age=(\d+)/i.exec(cacheControl);
  const seconds = match ? Number(match[1]) : NaN;
  return Number.isSafeInteger(seconds) && seconds > 0 ? seconds * 1000 : DEFAULT_JWKS_TTL_MS;
}

async function fetchKeys(): Promise<CachedKeys> {
  const response = await fetch(JWKS_URL, {
    headers: { Accept: 'application/json' },
    signal: AbortSignal.timeout(10_000),
  });

  if (!response.ok) {
    throw new Error(`Could not fetch Google signing keys: ${response.status}`);
  }

  const body = (await response.json()) as { keys?: JsonWebKey[] };
  const keys = new Map<string, KeyObject>();

  for (const jwk of body.keys ?? []) {
    // Only RSA signing keys are usable here, and a key with no id cannot be
    // selected by a token's header. Skipping rather than failing keeps one odd
    // entry in the set from taking down every sign-in.
    if (!jwk.kid || jwk.kty !== 'RSA' || (jwk.alg && jwk.alg !== 'RS256')) continue;
    try {
      keys.set(jwk.kid, createPublicKey({ key: jwk as never, format: 'jwk' }));
    } catch (err) {
      logger.warn({ err, kid: jwk.kid }, 'Skipped an unreadable Google signing key');
    }
  }

  if (keys.size === 0) {
    throw new Error('Google signing key set contained no usable RSA keys');
  }

  return { keys, expiresAt: Date.now() + maxAgeMs(response.headers.get('cache-control')) };
}

/**
 * The current key set, fetched at most once at a time.
 *
 * Google rotates these, so a `kid` that misses is a reason to refetch rather
 * than to reject — but only once per call, or an invalid token becomes a way to
 * make this server hammer Google.
 */
async function keySet(forceRefresh = false): Promise<CachedKeys> {
  if (!forceRefresh && cache && cache.expiresAt > Date.now()) return cache;
  if (!forceRefresh && inFlight) return inFlight;

  inFlight = fetchKeys()
    .then(fetched => {
      cache = fetched;
      return fetched;
    })
    .finally(() => {
      inFlight = undefined;
    });

  return inFlight;
}

function decodeSegment(segment: string): unknown {
  return JSON.parse(Buffer.from(segment, 'base64url').toString('utf8')) as unknown;
}

interface IdTokenClaims {
  iss?: string;
  aud?: string;
  sub?: string;
  email?: string;
  email_verified?: boolean | string;
  name?: string;
  exp?: number;
  iat?: number;
  nbf?: number;
}

/**
 * Verifies a Google ID token, or returns null.
 *
 * Every rejection returns the same thing for the same reason the client-side
 * decoder does: the caller's response to "this token is not good" does not vary
 * with why, and a response that varies is a response that teaches an attacker
 * something. The reason is logged, not returned.
 */
export async function verifyGoogleIdToken(
  token: string,
  now = Date.now(),
): Promise<VerifiedGoogleUser | null> {
  const audience = googleClientId();
  if (!audience) return null;

  const segments = token.split('.');
  if (segments.length !== 3) return null;

  let header: { alg?: string; kid?: string };
  let claims: IdTokenClaims;
  try {
    header = decodeSegment(segments[0]) as { alg?: string; kid?: string };
    claims = decodeSegment(segments[1]) as IdTokenClaims;
  } catch {
    return null;
  }

  // Pinned, not read from the token: accepting whatever `alg` the token names
  // is how "none" and key-confusion attacks get in.
  if (header.alg !== 'RS256' || !header.kid) return null;

  const signingInput = Buffer.from(`${segments[0]}.${segments[1]}`, 'utf8');
  const signature = Buffer.from(segments[2], 'base64url');

  let keys = await keySet();
  let key = keys.keys.get(header.kid);
  if (!key) {
    // Unknown key id: Google may have rotated since the set was cached.
    keys = await keySet(true);
    key = keys.keys.get(header.kid);
  }
  if (!key) return null;

  if (!verifySignature('sha256', signingInput, key, signature)) return null;

  // Only past this line do the claims mean anything.
  if (!claims.iss || !VALID_ISSUERS.has(claims.iss)) return null;
  if (claims.aud !== audience) return null;
  if (typeof claims.exp !== 'number') return null;

  const nowSeconds = now / 1000;
  if (claims.exp + CLOCK_SKEW_SECONDS <= nowSeconds) return null;
  if (typeof claims.iat === 'number' && claims.iat - CLOCK_SKEW_SECONDS > nowSeconds) return null;
  if (typeof claims.nbf === 'number' && claims.nbf - CLOCK_SKEW_SECONDS > nowSeconds) return null;

  // An unverified address is one the account holder never proved they own, so
  // it cannot be what a purchase is matched against.
  const emailVerified = claims.email_verified === true || claims.email_verified === 'true';
  if (!claims.sub || !claims.email || !emailVerified) return null;

  return {
    subject: claims.sub,
    email: claims.email.trim().toLowerCase(),
    name: claims.name ?? claims.email,
    expiresAt: claims.exp * 1000,
  };
}

function bearerToken(req: Request): string {
  const match = /^Bearer[ ]+(.+)$/i.exec((req.get('authorization') ?? '').trim());
  return match ? match[1].trim() : '';
}

/**
 * The verified caller, or null when there is no usable token.
 *
 * For routes that behave differently for a signed-in visitor but still work
 * without one — checkout being the case that matters, since buying the course
 * must not require an account.
 */
export async function verifiedGoogleUser(req: Request): Promise<VerifiedGoogleUser | null> {
  const token = bearerToken(req);
  if (!token) return null;
  try {
    return await verifyGoogleIdToken(token);
  } catch (err) {
    // A network failure reaching Google is not the caller's fault, but it is
    // also not proof of identity: treat it as signed out and say so in the log.
    logger.warn({ err, requestId: req.id }, 'Could not verify a Google ID token');
    return null;
  }
}

/** Express typing for a request that has been through `requireGoogleUser`. */
export interface AuthenticatedRequest extends Request {
  googleUser: VerifiedGoogleUser;
}

/** Rejects the request unless it carries a verified Google ID token. */
export async function requireGoogleUser(
  req: Request,
  res: Response,
  next: NextFunction,
): Promise<void> {
  if (!isGoogleAuthConfigured()) {
    res.status(503).json({ error: 'Sign-in is not configured on this server.' });
    return;
  }

  const user = await verifiedGoogleUser(req);
  if (!user) {
    res.status(401).json({ error: 'A valid Google ID token is required.' });
    return;
  }

  (req as AuthenticatedRequest).googleUser = user;
  next();
}

File: server/app/google_auth.py
from __future__ import annotations

import asyncio
import base64
import json
import logging
import os
import time
from dataclasses import dataclass
from typing import Any

import httpx
from cryptography.hazmat.primitives.asymmetric import padding, rsa
from cryptography.hazmat.primitives.hashes import SHA256

JWKS_URL = "https://www.googleapis.com/oauth2/v3/certs"
VALID_ISSUERS = {"accounts.google.com", "https://accounts.google.com"}
CLOCK_SKEW_SECONDS = 60
DEFAULT_JWKS_TTL_SECONDS = 3600
logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class GoogleUser:
    subject: str
    email: str
    name: str
    picture: str
    expires_at: int


_keys: dict[str, rsa.RSAPublicKey] = {}
_keys_expire_at = 0.0
_keys_lock = asyncio.Lock()


def _decode(segment: str) -> dict[str, Any]:
    padded = segment + "=" * (-len(segment) % 4)
    value = json.loads(base64.urlsafe_b64decode(padded).decode("utf-8"))
    if not isinstance(value, dict):
        raise ValueError("JWT segment is not an object")
    return value


def _max_age(headers: httpx.Headers) -> int:
    for part in headers.get("cache-control", "").split(","):
        name, _, value = part.strip().partition("=")
        if name.lower() == "max-age" and value.isdigit() and int(value) > 0:
            return int(value)
    return DEFAULT_JWKS_TTL_SECONDS


async def _fetch_keys() -> tuple[dict[str, rsa.RSAPublicKey], float]:
    async with httpx.AsyncClient(timeout=10) as client:
        response = await client.get(JWKS_URL, headers={"accept": "application/json"})
        response.raise_for_status()
    parsed: dict[str, rsa.RSAPublicKey] = {}
    for jwk in response.json().get("keys", []):
        if (
            not isinstance(jwk, dict)
            or not jwk.get("kid")
            or jwk.get("kty") != "RSA"
            or jwk.get("alg", "RS256") != "RS256"
        ):
            continue
        try:
            exponent = int.from_bytes(base64.urlsafe_b64decode(jwk["e"] + "=="))
            modulus = int.from_bytes(base64.urlsafe_b64decode(jwk["n"] + "=="))
            parsed[jwk["kid"]] = rsa.RSAPublicNumbers(exponent, modulus).public_key()
        except (KeyError, TypeError, ValueError):
            logger.warning("Skipped an unreadable Google signing key")
    if not parsed:
        raise ValueError("Google signing key set contained no usable RSA keys")
    return parsed, time.time() + _max_age(response.headers)


async def _key_set(force: bool = False) -> dict[str, rsa.RSAPublicKey]:
    global _keys, _keys_expire_at
    if not force and _keys and _keys_expire_at > time.time():
        return _keys
    async with _keys_lock:
        if not force and _keys and _keys_expire_at > time.time():
            return _keys
        _keys, _keys_expire_at = await _fetch_keys()
        return _keys


def reset_key_cache() -> None:
    global _keys, _keys_expire_at
    _keys = {}
    _keys_expire_at = 0


async def verify_google_id_token(token: str, now: float | None = None) -> GoogleUser | None:
    audience = os.getenv("GOOGLE_CLIENT_ID", "").strip()
    if not audience:
        return None
    segments = token.split(".")
    if len(segments) != 3:
        return None
    try:
        header, claims = _decode(segments[0]), _decode(segments[1])
        if header.get("alg") != "RS256" or not isinstance(header.get("kid"), str):
            return None
        keys = await _key_set()
        key = keys.get(header["kid"])
        if key is None:
            keys = await _key_set(force=True)
            key = keys.get(header["kid"])
        if key is None:
            return None
        signature = base64.urlsafe_b64decode(segments[2] + "=" * (-len(segments[2]) % 4))
        key.verify(signature, f"{segments[0]}.{segments[1]}".encode(), padding.PKCS1v15(), SHA256())
    except Exception:
        return None

    current = now if now is not None else time.time()
    exp, issued, not_before = claims.get("exp"), claims.get("iat"), claims.get("nbf")
    verified = claims.get("email_verified") in (True, "true")
    if (
        claims.get("iss") not in VALID_ISSUERS
        or claims.get("aud") != audience
        or not isinstance(exp, (int, float))
        or exp + CLOCK_SKEW_SECONDS <= current
        or isinstance(issued, (int, float))
        and issued - CLOCK_SKEW_SECONDS > current
        or isinstance(not_before, (int, float))
        and not_before - CLOCK_SKEW_SECONDS > current
        or not isinstance(claims.get("sub"), str)
        or not isinstance(claims.get("email"), str)
        or not verified
    ):
        return None
    email = claims["email"].strip().lower()
    return GoogleUser(
        subject=claims["sub"],
        email=email,
        name=claims.get("name") if isinstance(claims.get("name"), str) else email,
        picture=claims.get("picture") if isinstance(claims.get("picture"), str) else "",
        expires_at=int(exp),
    )
